power_grid_model_c/power_grid_model_cpp/include/power_grid_model_cpp/basics.hpp

```diff
@@ -0,0 +1,35 @@
+// SPDX-FileCopyrightText: Contributors to the Power Grid Model project <powergridmodel@lfenergy.org>
+//
+// SPDX-License-Identifier: MPL-2.0
+
+#pragma once
+#ifndef POWER_GRID_MODEL_CPP_BASICS_HPP
+#define POWER_GRID_MODEL_CPP_BASICS_HPP
+
+#ifdef PGM_DLL_EXPORTS
+#error "You should not define PGM_DLL_EXPORTS when using the C++ headers"
+#endif
+
+#include <exception>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "basics.h"
+
+namespace power_grid_model_cpp {
+
+using Idx = PGM_Idx;
+using ID = PGM_ID;
+
+// custom deleter
+template <auto func> struct DeleterFunctor {
+    template <typename T> void operator()(T* arg) const { func(arg); }
+};
+
+// unique pointer definition
+template <typename T, auto func> using UniquePtr = std::unique_ptr<T, DeleterFunctor<func>>;
```

Comment:
maybe `namespace detail`

Reply:
I had it implemented like that before, but since `UniquePtr` and `DeleteFunctor` now live within `basics.hpp`, should we still add `namespace detail`? Is it to add another layer for the possible user since these are used for the "internal" stuff of the wrapper?

Reply:
indeed. this is internally used and not really necessary for the user to even know it exists.